pkg/operator/ceph/cluster/crash/add.go

```diff
@@ -0,0 +1,155 @@
+
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return fmt.Errorf("failed to create a new %q. %+v", controllerName, err)
+	}
+
+	// Watch for changes to the nodes
+	specChangePredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			nodeOld := e.ObjectOld.DeepCopyObject().(*corev1.Node)
+			nodeNew := e.ObjectNew.DeepCopyObject().(*corev1.Node)
+			return !reflect.DeepEqual(nodeOld.Spec, nodeNew.Spec)
+		},
+	}
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{}, specChangePredicate)
+	if err != nil {
+		return fmt.Errorf("failed to watch for node changes. %+v", err)
+	}
+
+	// Watch for changes to the ceph-crash deployments
+	err = c.Watch(
+		&source.Kind{Type: &appsv1.Deployment{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: handler.ToRequestsFunc(func(obj handler.MapObject) []reconcile.Request {
+				deployment, ok := obj.Object.(*appsv1.Deployment)
+				if !ok {
+					return []reconcile.Request{}
+				}
+				labels := deployment.GetLabels()
+				if appName, ok := labels[k8sutil.AppAttr]; !ok || appName != AppName {
+					return []reconcile.Request{}
+				}
+				nodeName, ok := deployment.Spec.Template.ObjectMeta.Labels[NodeNameLabel]
+				if !ok {
+					return []reconcile.Request{}
+				}
+				req := reconcile.Request{NamespacedName: types.NamespacedName{Name: nodeName}}
+				return []reconcile.Request{req}
+			}),
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to watch for changes on the ceph-crash deployment. %+v", err)
+	}
+
+	// Watch for changes to the ceph pod nodename and enqueue their nodes
+	err = c.Watch(
+		&source.Kind{Type: &corev1.Pod{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: handler.ToRequestsFunc(func(obj handler.MapObject) []reconcile.Request {
+				pod, ok := obj.Object.(*corev1.Pod)
+				if !ok {
+					return []reconcile.Request{}
+				}
+				nodeName := pod.Spec.NodeName
+				if len(nodeName) < 1 {
+					return []reconcile.Request{}
+				}
+				if isCephPod(pod.Name) {
+					req := reconcile.Request{NamespacedName: types.NamespacedName{Name: nodeName}}
+					return []reconcile.Request{req}
+				}
+				return []reconcile.Request{}
+			}),
+		},
+		// only enqueue the update event if the pod moved nodes
+		predicate.Funcs{
+			UpdateFunc: func(event event.UpdateEvent) bool {
+				oldPod, ok := event.ObjectOld.(*corev1.Pod)
+				if !ok {
+					return false
+				}
+				newPod, ok := event.ObjectNew.(*corev1.Pod)
+				if !ok {
+					return false
+				}
+				// only enqueue if the nodename has changed
+				if oldPod.Spec.NodeName == newPod.Spec.NodeName {
+					return false
+				}
+				return true
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to watch for changes on the ceph pod nodename and enqueue their nodes. %+v", err)
+	}
+
+	return nil
+}
+
+func isCephPod(podName string) bool {
```

Comment:
This handler will receive non-rook pods as well, can we use label checking instead of podName checking here?

I see that when listing the ceph pods you are using the labels, so this will have no impact on functionality other than pods that happen to have this string will trigger an unnecessary reconcile loop. (eg a pod named mongodb would trigger a reconcile loop).

Reply:
Good catch!